exchange-proxy/features/meta-transactions.md

```diff
@@ -0,0 +1,212 @@
+
+## Design
+
+The V3 Exchange simply set a `currentContextAddress` storage variable to the meta-transaction owner then `delegatecall`ed the calldata into itself.
+
+It would be pretty risky for the Exchange Proxy to `delegatecall` user-supplied calldata because there are several “internal” functions registered to the Exchange Proxy that are guarded by an `onlySelf` modifier (`msg.sender == address(this)`). So some kind of white-listed would be required.
+
+Also, Exchange Proxy features are encouraged to expose an internal variant of functions that allow the context address to be passed explicitly. This removes the need to write to storage during a meta-transaction call. However, the calldata is no longer what we actually execute, as we have to decode it and manually call the internal variant, which likely takes different parameters.
+
+### Features
+
+Two new feature contracts will be added to the Exchange Proxy:
+
+* `MetaTransactions`
+    * Implements and exposes functions for working with meta-transactions.
+        * `executeMetaTransaction()` Execute a single meta-transaction.
+        * `batchExecuteMetaTransactions()` Execute multiple meta-transactions in sequence.
+        * `_executeMetaTransaction()`Internal variant of `executeMetaTransaction()` that allows the sender to be explicitly provided.
+        * `getMetaTransactionHash()` Return the hash of the meta-transaction.
+        * `getMetaTransactionExecutedBlock()` Get the block height at which a meta-transaction was executed (0 if never executed).
+        * `getMetaTransactionHashExecutedBlock()` Get the block height at which a meta-transaction *has* was executed (0 if never executed).
+    * Owns the storage for tracking already executed transactions.
+    * Reentrancy is allowed on the execute functions.
+        * Reentrancy guards should be placed on the functions being called if necessary.
+* `SignatureValidator`
+    * Implements and exposes functions for validating signatures.
+    * This allows us to support new signature types across all features without having to redeploy them.
+    * Initially will just support `eth_sign` and EIP712 signatures.
+
+
+![meta-transactions](https://raw.githubusercontent.com/0xProject/0x-protocol-specification/master/exchange-proxy/img/meta-transactions.png)
+
+## The MetaTransactions Feature
+
+### Interface
+```
+interface IMetaTransactions {
+
+    struct MetaTransactionData {
+        // Signer of meta-transaction. On whose behalf we execute the MTX.
+        address signer;
+        // Allowed sender, or 0 for anyone.
+        address sender;
+        // Minimum gas price.
+        uint256 minGasPrice;
+        // Maximum gas price.
+        uint256 maxGasPrice;
+        // MTX is invalid after this time.
+        uint256 expirationTimeSeconds;
+        // Nonce to make this MTX unique.
+        uint256 salt;
+        // Call data to execute (but not really).
+        bytes callData;
```

Comment:
Let's include how `callData` should be encoded. Maybe just have a struct below like:
```
struct CallData {
    IERC20Token inputToken,
    IERC20Token outputToken,
    uint256 inputTokenAmount,
    uint256 minOutputTokenAmount,
    Transformation[] transformations
}
```